pkg/console/operator/health.go

```diff
@@ -0,0 +1,85 @@
+	"os"
+	"time"
+
+	deploymentv1 "k8s.io/api/apps/v1"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	routev1 "github.com/openshift/api/route/v1"
+
+	"github.com/openshift/console-operator/pkg/console/status"
+	deploymentsub "github.com/openshift/console-operator/pkg/console/subresource/deployment"
+	routesub "github.com/openshift/console-operator/pkg/console/subresource/route"
+)
+
+func (co *consoleOperator) CheckRouteHealth(opConfig *operatorv1.Console, rt *routev1.Route) {
+	status.HandleDegraded(func() (conf *operatorv1.Console, prefix string, reason string, err error) {
+		prefix = "RouteHealth"
+		client := http.Client{Timeout: time.Duration(5 * time.Second)}
+		req, err := http.NewRequest(http.MethodHead, "https://"+rt.Spec.Host+"/health", nil)
+		if err != nil {
+			return opConfig, prefix, "FailedRequest", fmt.Errorf("failed to build request to route: %v", err)
+		}
+		resp, err := client.Do(req)
+		if err != nil {
+			return opConfig, prefix, "FailedGet", fmt.Errorf("failed to GET route: %v", err)
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != http.StatusOK {
+			return opConfig, prefix, "StatusError", fmt.Errorf("route not yet available, /health returns '%s'", resp.Status)
+
+		}
+		return opConfig, prefix, "", nil
+	}())
+
+	status.HandleAvailable(opConfig, "Route", "FailedAdmittedIngress", func() error {
+		if !routesub.IsAdmitted(rt) {
+			return errors.New("console route is not admitted")
+		}
+		return nil
+	}())
+}
+
+func (co *consoleOperator) CheckDeploymentHealth(opConfig *operatorv1.Console, deployment *deploymentv1.Deployment, toUpdate bool) {
+	status.HandleAvailable(func() (conf *operatorv1.Console, prefix string, reason string, err error) {
+		prefix = "Deployment"
+		if !deploymentsub.IsReady(deployment) {
+			return opConfig, prefix, "InsufficientReplicas", errors.New(fmt.Sprintf("%v pods available for console deployment", deployment.Status.ReadyReplicas))
+		}
+		if !deploymentsub.IsReadyAndUpdated(deployment) {
+			return opConfig, prefix, "FailedUpdate", errors.New(fmt.Sprintf("%v replicas ready at version %s", deployment.Status.ReadyReplicas, os.Getenv("RELEASE_VERSION")))
+		}
+		return opConfig, prefix, "", nil
+	}())
+
+	status.HandleAvailable(func() (conf *operatorv1.Console, prefix string, reason string, err error) {
+		prefix = "Deployment"
+		if !deploymentsub.IsReady(deployment) {
+			return opConfig, prefix, "InsufficientReplicas", errors.New(fmt.Sprintf("%v pods available for console deployment", deployment.Status.ReadyReplicas))
+		}
+		if !deploymentsub.IsReadyAndUpdated(deployment) {
+			return opConfig, prefix, "FailedUpdate", errors.New(fmt.Sprintf("%v replicas ready at version %s", deployment.Status.ReadyReplicas, os.Getenv("RELEASE_VERSION")))
+		}
+		return opConfig, prefix, "", nil
+	}())
+
+	status.HandleProgressing(opConfig, "SyncLoopRefresh", "InProgress", func() error {
+		if toUpdate {
+			return errors.New("Changes made during sync updates, additional sync expected.")
```

Comment:
Thinking if it wouldn't make sense to also mention the resource, just to have easier debugging.

```suggestion
			return errors.New("Deployment changes made during sync updates, additional sync expected.")
```

Reply:
The `toUpdate` bool though could be any of the resources, it may be something else (that may or may not eventually cause the deployment to roll)

Reply:
That first `if toUpdate` is a little bit unfortunate in that way, but hopefully all of our other statuses are clear enough that we indicate what is going on.